File: workflows/tools/options.rb
#!/usr/bin/env ruby

require 'fileutils'
require 'optparse'
require 'pathname'
require 'tempfile'
require 'yaml'

command = ARGV.shift
settings = {}
OptionParser.new do |parser|
  parser.on('--work-root PATH', '本轮工作根目录的绝对路径') { |value| settings[:work_root] = value }
end.parse!(ARGV)

abort '用法: ruby options.rb set|get --work-root /absolute/path 选项名 [值]' unless %w[set get].include?(command)
abort '请提供绝对工作根目录' unless settings[:work_root] && Pathname.new(settings[:work_root]).absolute?

name = ARGV.shift
abort '请提供选项名' if name.nil? || name.empty?

schema_path = File.expand_path('../流程.yaml', __dir__)
schema = YAML.safe_load(File.read(schema_path)) || {}
definition = (schema['options'] || {})[name]
abort "未在流程.yaml中声明选项: #{name}" unless definition.is_a?(Hash)

root = settings[:work_root]
path = File.join(root, 'options.yaml')

if command == 'get'
  abort 'options.yaml尚未创建' unless File.file?(path)
  options = YAML.safe_load(File.read(path)) || {}
  abort 'options.yaml必须是键值映射' unless options.is_a?(Hash)
  abort "选项尚未设置: #{name}" unless options.key?(name)
  puts "#{name}: #{options[name].inspect}"
  exit 0
end

raw = ARGV.shift
abort '请提供选项值' if raw.nil? || !ARGV.empty?

value = case definition['type']
        when 'boolean'
          abort '布尔选项只接受true或false' unless %w[true false].include?(raw)
          raw == 'true'
        when 'string'
          raw
        when 'integer'
          abort '整数选项格式无效' unless raw.match?(/\A-?\d+\z/)
          raw.to_i
        when 'enum'
          abort '选项值不在允许列表中' unless Array(definition['values']).include?(raw)
          raw
        else
          abort "不支持的选项类型: #{definition['type']}"
        end

FileUtils.mkdir_p(root)
File.open(File.join(root, '.options.yaml.lock'), File::RDWR | File::CREAT, 0600) do |lock|
  lock.flock(File::LOCK_EX)
  options = File.file?(path) ? YAML.safe_load(File.read(path)) : {}
  options ||= {}
  abort 'options.yaml必须是键值映射' unless options.is_a?(Hash)
  options[name] = value

  Tempfile.create(['.options-', '.yaml'], root) do |temp|
    temp.write(YAML.dump(options))
    temp.flush
    temp.fsync
    File.rename(temp.path, path)
  end
end

puts "#{name}: #{value.inspect}"

File: workflows/tools/options.py
#!/usr/bin/env python3
"""Read/write flat workflow options using only the Python 3 standard library."""

import argparse
from contextlib import contextmanager
import json
import os
from pathlib import Path
import re
import sys
import tempfile
import time


def valid_name(name):
    return isinstance(name, str) and re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", name)


def unique_object(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError("Duplicate option: " + key)
        result[key] = value
    return result


def read_options(path):
    with path.open(encoding="utf-8-sig") as source:
        options = json.load(source, object_pairs_hook=unique_object)
    if not isinstance(options, dict):
        raise ValueError("options.json must contain a flat object.")
    for key, value in options.items():
        if not valid_name(key) or type(value) not in (bool, int, str):
            raise ValueError("Options require valid names and boolean, integer or string values.")
    return options


def parse_value(raw, value_type):
    if value_type == "boolean":
        if raw not in ("true", "false"):
            raise ValueError("Boolean values must be true or false.")
        return raw == "true"
    if value_type == "integer":
        if not re.fullmatch(r"-?[0-9]+", raw):
            raise ValueError("Invalid integer value.")
        return int(raw)
    return raw


@contextmanager
def write_lock(root):
    lock = root / ".options.json.lock"
    deadline = time.monotonic() + 10
    while True:
        try:
            descriptor = os.open(str(lock), os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
            break
        except FileExistsError:
            if time.monotonic() >= deadline:
                raise ValueError("Cannot acquire .options.json.lock; check active writers.")
            time.sleep(0.1)
    try:
        os.close(descriptor)
        yield
    finally:
        lock.unlink()


def write_options(path, options):
    temporary = None
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", newline="\n", dir=str(path.parent),
            prefix=".options-", suffix=".json", delete=False
        ) as target:
            temporary = Path(target.name)
            json.dump(options, target, ensure_ascii=False, indent=2, allow_nan=False)
            target.write("\n")
            target.flush()
            os.fsync(target.fileno())
        os.replace(str(temporary), str(path))
        temporary = None
    finally:
        if temporary is not None:
            temporary.unlink()


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    commands = parser.add_subparsers(dest="command", required=True)
    for command in ("set", "get"):
        sub = commands.add_parser(command)
        sub.add_argument("--work-root", required=True, type=Path)
        sub.add_argument("name")
        if command == "set":
            sub.add_argument("value")
            sub.add_argument("--type", choices=("boolean", "integer", "string"), default="boolean")
    args = parser.parse_args()
    if not args.work_root.is_absolute():
        parser.error("--work-root must be an absolute path.")
    if not valid_name(args.name):
        parser.error("Invalid option name.")

    path = args.work_root / "options.json"
    if args.command == "get":
        options = read_options(path)
        if args.name not in options:
            raise ValueError("Option is not set: " + args.name)
        value = options[args.name]
    else:
        value = parse_value(args.value, args.type)
        args.work_root.mkdir(parents=True, exist_ok=True)
        with write_lock(args.work_root):
            options = read_options(path) if path.exists() else {}
            options[args.name] = value
            write_options(path, options)
    print(json.dumps({args.name: value}, ensure_ascii=True))


if __name__ == "__main__":
    try:
        main()
    except (OSError, ValueError) as error:
        print(str(error), file=sys.stderr)
        sys.exit(1)
